extensions/2.0/Vendor/BENTLEY_primitive_restart/README.md

```diff
@@ -0,0 +1,154 @@
+
+```
+[0, 1, 255, 2, 3, 4]
+[0, 1]     [2, 3, 4]
+```
+
+This permits two equivalent representations of the geometry without duplicating any binary data. In glTF, the accessors look like this:
+
+```json
+    "accessors": [
+        {
+            "bufferView": 0,
+            "count": 2,
+            "componentType": 5121,
+            "type": "SCALAR",
+            "name": "Line string 1 indices"
+        },
+        {
+            "bufferView": 0,
+            "byteOffset": 3,
+            "count": 3,
+            "componentType": 5121,
+            "type": "SCALAR",
+            "name": "Line string 2 indices"
+        },
+        {
+            "bufferView": 0,
+            "componentType": 5121,
+            "count": 6,
+            "type": "SCALAR",
+            "name": "All indices"
+        },
+        {
+            "bufferView": 1,
+            "componentType": 5126,
+            "count": 5,
+            "type": "VEC3",
+            "max": [
+                0.5,
+                0.5,
+                0.0
+            ],
+            "min": [
+                -0.5,
+                -0.5,
+                0.0
+            ],
+            "name": "Positions accessor"
+        }
+    ],
+
+```
+
+The mesh looks like this:
+
+```json
+    "meshes": [
+        {
+            "primitives": [
+                {
+                    "attributes": {
+                        "POSITION": 0
+                    },
+                    "indices": 0,
+                    "material": 0,
+                    "mode": 3
+                },
+                {
+                    "attributes": {
+                        "POSITION": 0
+                    },
+                    "indices": 1,
+                    "material": 0,
+                    "mode": 3
+                }
+            ],
+            "extensions": {
+                "BENTLEY_primitive_restart": {
+                    "primitiveGroups": [
+                        {
+                            "primitives": [
+                                0,
+                                1
+                            ],
+                            "indices": 2
+                        }
+                    ]
+                }
+            }
+        }
+    ],
+```
+
+By default, this mesh draws two separate line strip primitives, each using its own `indices` accessor. The `BENTLEY_primitive_restart` extension specifies that both primitives can be replaced with a single one using a combined `indices` accessor containing primitive restart values.
+
+## glTF Schema Updates
+
+The `BENTLEY_primitive_restart` extension is applied to a mesh. Its `primitiveGroups` property is a list of groups of primitives that can be replaced with a single primitive using primitive restart. Each group is described by a list of indices into `mesh.primitives`, along with the index of the accessor that supplies the vertex indices for the replacement primitive.
+
+## Constraints
```

Comment:
There is a high-level issue with all these constraints.

Let's say an asset that violates them (but perfectly valid otherwise) is loaded in an engine that does not support the extension. In this case, the asset would render fine.

But if the same asset is loaded in an engine that does support the extension, its rendering becomes undefined because the base spec does not generally prescribe what to do with invalid content.

Since this extension is fundamentally a performance hint and does not have any user-observable effects, it could have a built-in fallback to avoid undefined behavior by saying that if these constraints are violated, the extension becomes implicitly disabled and the affected primitive group (or the whole mesh) is rendered as usual.

Reply:
Added language to this effect.

Reply:
@lexaknyazev *requiring* clients to implement fallback amounts to requiring them to exhaustively validate all of the constraints. Some of that may be trivial (e.g., is `primitiveGroups` empty), but not all of it (e.g., do all primitives in a group have exactly the same property values recursively). Fallback is nice to have, but the onus should be on the encoders to conform to the spec.